# **Lab 3: Using Objects and Good Programming Practices**
---

### **Description**
In today's lab, you will learn how to work with Attributes and Methods in Python. You will also learn how to interpret and correct coding errors, and how to search for and navigate documentation.

<br>

### **Lab Structure**
**Part 1**: [Review](#p1)

**Part 2**: [Attributes and Methods](#p2)

**Part 3**: [Debugging](#p3)

**Part 4**: [Additional Practice [OPTIONAL]](#p4)
>
> **Part 4.1**: [The Fundamentals](#p4.1)
>
> **Part 4.2**: [Coding Challenges](#p4.2)

**Part 5**: [The `print` Function [OPTIONAL]](#p5)

<br>


### **Learning Objectives**
By the end of this lab, we will:
* Understand how to interact with objects using methods and attributes.
* Improve your debugging skills by correcting common coding errors.
* Understand how to search for documentation in various libraries.

<br>

###**Resources**
* [Python Basics Cheat Sheet](https://docs.google.com/document/d/1uLao-xKh3tYIiCYyiqf7oas4qTlr5XYKQ0QpzX_Xzag/edit?usp=sharing)


<br>


**Before starting, run the code below to import all necessary functions and libraries.**


In [ ]:
# @title

import warnings
import sympy
import numpy
import matplotlib
import time
import math
from matplotlib import pyplot

warnings.filterwarnings("ignore")

class baby():
  def __init__(self,age=0, mood="crying", hunger="hungry"):
    self.age = age
    self.mood = mood
    self.hunger = hunger
  def feed(self):
    self.hunger = "full"
  def play(self):
    self.mood = "happy"

class second_lab():
  def __init__(self,TA="NO TA SET"):
    self.topic = "Introduction to Objects and Debugging"
    self.attendance_code = 1234
    self.attendees = []
    self.TA = TA
  def check_in(self,name):
    self.attendees.append(name)
  def lab_summary(self):
    print("Today in lab, we had an " + self.topic + " with our TA, " + self.TA +". We had some great people in attendance, such as:" + ", ".join(self.attendees)+". The attendance code is " + str(self.attendance_code)+".")

def calculate_area(length, width):
  area = length * width
  return area

class plant():
  def __init__(self):
    self.planted = 0
    self.day_of_last_growth = 0
    self.day = 0
    self.water = 0
    self.light = 0
    self.health = 10
    self.stage = 0
    self.perfect_light = numpy.random.choice([1,2,3,4,5,6,7,8,9,10])
    self.perfect_water = numpy.random.choice([1,2,3,4,5,6,7,8,9,10])
  def water_plant(self,amount):
    if amount > 0:
      print("You watered your plant.")
      self.water += round(amount)
    elif amount < 0:
      raise ValueError("You can't remove water from your plant! You will have to wait a bit for it to dry out.")
    else:
      print("You decide to not water your plant.")
  def set_light(self,level):
    if level <0:
      raise ValueError("You can't set the light level to negative values.")
    elif level > 10:
      raise ValueError("Your light is not that strong")
    else:
      self.light = round(level)
      print("You set the light to level " + str(self.light))

  def check_health(self):
    print("Your plant's health is a " + str(self.health)+ "/" + str(10+(5*self.stage))+".")
    print("If you had to rate the wetness of the soil, it would be a " + str(self.water)+ "/10.")
    print("Your light is turned up to " + str(self.light)+ ".")
    if self.stage == 10:
      print("Your plant is full grown.")
  def wait(self,num_days):
    for i in range(num_days):
      self.day += 1
      print('Day ' + str(self.day))
      if self.perfect_light > self.light:
        print("Your plant needs more light")
        self.health += -1*abs(self.perfect_light- self.light)/10
      elif self.perfect_light < self.light:
        print("Your plant needs less light")
        self.health += -1*abs(self.perfect_light- self.light)/10
      else:
        print("Your plant looks happy with the light it has.")
      if self.perfect_water > self.water:
        print("Your plant needs more water")
        self.health += -1*abs(self.perfect_water- self.water)/10
      elif self.perfect_water < self.water:
        print("Your plant needs less water")
        self.health += -1*abs(self.perfect_water- self.water)/10
      else:
        print("Your plant looks happy with the water it had.")
      if self.perfect_light == self.light and self.perfect_water == self.water:
        self.health = (self.health + 10+(5*self.stage))/2
        if self.day- self.day_of_last_growth >= 5 and self.stage < 10:
          print("Your plant grew a bit! Its preferances may have changed.")
          self.stage += 1
          self.health += 5
          self.day_of_last_growth = self.day
          self.perfect_light = numpy.random.choice([1,2,3,4,5,6,7,8,9,10])
          self.perfect_water = numpy.random.choice([1,2,3,4,5,6,7,8,9,10])
          if self.stage == 10:
            print("Congrats! Your plant is full grown.")
      self.water += -1*self.light
      if self.water < 0:
        self.water = 0
      if self.health <= 0:
        print("Your plant died! You decide to plant a new seed. Treat this one better!")
        self.planted = 0
        self.day_of_last_growth = 0
        self.day = 0
        self.water = 0
        self.light = 0
        self.health = 10
        self.stage = 0
        self.perfect_light = numpy.random.choice([1,2,3,4,5,6,7,8,9,10])
        self.perfect_water = numpy.random.choice([1,2,3,4,5,6,7,8,9,10])

def dynamic_find_roots(a,b,c):
  if b**2-4*a*c == 0:
    return -b/(2*a)
  elif b**2 - 4*a*c <0:
    return (-b+1j*(4*a*c-b**2)**0.5)/(2*a),(-b-1j*(4*a*c-b**2)**0.5)/(2*a)
  elif b**2 - 4*a*c >0:
    return (-b+(b**2-4*a*c)**0.5)/(2*a),(-b-(b**2-4*a*c)**0.5)/(2*a)
print("Libraries imported successfully!")

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 14.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 26.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 5.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 12.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 5.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.5/37.5 MB 38.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 112.7/112.7 kB 10.3 MB/s eta 0:00:00
Libraries imported successfully!


<a name="p1"></a>

---
## **Part 1: Review**
---

#### **Problem #1.1**

**Together**, let's label the following data types.

<br>

**NOTE**: You can change the text below by double clicking on the cell.

```
2
8.9
"Quantum is Cool!"
5.0
["string"]
```

#### **Problem #1.2**

**Together**, let's write code below to calculate 6 times 30 minus 15.5.

In [ ]:
# COMPLETE THIS CODE

#### **Problem #1.3**

**Together**, let's write code below to create a variables called `num_qubits` that equals 100. Then output the value of the variable.

In [ ]:
# COMPLETE THIS CODE

#### **Problem #1.4**

**Independently**, write code below to create a variables called `num_gates` that equals 12. Then output the value of the variable.

In [ ]:
# COMPLETE THIS CODE

#### **Problem #1.5**

Write code below to create a variables called `circuit_size` that equals `num_qubits` times `num_gates`. Then output the value of the variable.

In [ ]:
# COMPLETE THIS CODE

<a name="p2"></a>

---
## **Part 2: Attributes and Methods**
---
Python and any libraries we install give us tons of objects to work with. In this part, we will practice interacting with objects using methods and attributes.

### **Problem #2.1**

**Together**, let's create a "baby" object and output its "age" attribute.


In [ ]:
jeremy = baby()
# COMPLETE THIS CODE

### **Problem #2.2**

**Together**, let's write code to set the baby's age to 2.


In [ ]:
# COMPLETE THIS CODE

### **Problem #2.3**

**Together**, output the baby's `hunger` attribute, then use its `feed` method (this method does not require any arguments). Finally, print out the new `hunger` attribute.


In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

###**Problem #2.4**

**Independently**, create a new `baby` object, and output the `mood` attribute. Then, use its `play` method and output the `mood` attribute once again.

In [ ]:
# COMPLETE THIS CODE

###**Problem #2.5**

**Independently**, create a new `baby` object. Output its age attribute. Next, use its `feed` method, use its `play` method, and set its age attribute to 1. Then, output its `mood`, `hunger`, and new `age` attributes.

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

### **Problem #2.6**

**Together**, create a `second_lab` object, and set the `TA` attribute to the name of your TA. Then, output the `TA` attribute.

In [ ]:
my_lab = # COMPLETE THIS CODE

### **Problem #2.7**

**Independently**, use the `attendance_code` attribute of the `second_lab` object to find today's attendance poll code.


In [ ]:
# COMPLETE THIS CODE

### **Problem #2.8**

**Independently**, use the `check_in` method of the `second_lab` object to add your name to the list of attendees. This method requires one argument: the name of the person checking in. Check that your name was added to the list of attendees by checking the `attendees` attribute.


In [ ]:
# COMPLETE THIS CODE

###**Problem #2.9**

**Independently**, create a `plant` object using its corresponding function. Then, using the following methods, check its health, water it, set it's light level, wait three days, and check its health again.

* `check_health()`
* `water_plant(amount)`
* `set_light(level)`
* `wait(num_days)`

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

### **Problem #2.10 [CHALLENGE]**
**Independently**, add the names of two of your classmates to the list of attendees. Then, check the [methods of the list object](https://www.w3schools.com/python/python_ref_list.asp) to find a method that will sort the list of attendees alphabetically. Finally, call the `lab_summary` method of the `second_lab` object to print out a lab summary (this method does not require any arguments)

**Note:** Whenever you get stuck on a piece of code, you can always look up the object or the function you are working with, and there will be documentation describing all methods, attriutes, and required arguments.

In [ ]:
# COMPLETE THIS CODE

---

<center>

### **Wait for Your Instructor to Continue**

---

<a name="p3"></a>

---
## **Part 3: Debugging**
---

In this part, we will explore how to identify and correct errors in Python.

<br>

**NOTE**: Here and in general, you may not be familiar with the syntax presented in every line of code. However, you can use what you *do know* to solve the problems here and, often, in general.

#### **Problem #3.1**

**Together,** remove the line of code that causes the error.

In [ ]:
my_list = [1,2,3,4]
print(my_list[1])
print(my_list[2])
print(my_list[3])
print(my_list[4])

#### **Problem #3.2**

**Together**, correctly print each entry from my_list.

In [ ]:
print(my_list[# COMPLETE THIS CODE

#### **Problem #3.3**

**Together,** debug the code shown below.

In [ ]:
#f(x) = ax^2 + bx + c
a = 1
b = 0
c = -9
x1 = (-b + (b**2 - 4*a*c)**0.5)/(2*a)
x1 = (-b - (b**2 - 4*a*c)**0.5)/(2*a)
print(x1)
print(X2)

3.0


NameError: ignored

#### **Problem #3.4**

**Independently,** debug the code shown below.

In [ ]:
length = "30 meters"
width = "10 meters"
area = calculate_area(length, width)
print("The area is " + area)

TypeError: ignored

#### **Problem #3.5**

**Independently,** debug the code shown below.

In [ ]:
my_name = "ricky"
my_name.Capitalize()
Print(my_name)

AttributeError: ignored

#### **Problem #3.6**

**Independently**, debug the code shown below.

In [ ]:
x = "47"
y = 3
z = x**y
print(z)

TypeError: ignored

#### **Problem #3.7**

**Independently**, debug the code shown below.

In [ ]:
print(willoughby.mood)
willoughby.play()
print(willoughby.mood)

NameError: ignored

#### **Problem #3.8**

**Independently**, debug the code shown below.

In [ ]:
cactus = plant
cactus.check_health

<function __main__.plant.check_health(self)>

In [ ]:
cactus.water_plant("1")
cactus.set_light("8")
cactus.check_health()

TypeError: ignored

In [ ]:
cactus.wait(3)
print(cactus.check_health)

TypeError: ignored

<a name="p4"></a>

---
## **Part 4: Additional Practice [OPTIONAL]**
---

<a name="p4.1"></a>

---
### **Part 4.1: The Fundamentals**
---

This part contains questions on the basic syntax that we have learned. Use this section to practice and check your understanding.

#### **Problem #4.1.1**

Write code that outputs the names of three different countries. **HINT**: You will need a new code cell for every output.

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

#### **Problem #4.1.2**

Write code that outputs the result of multiplying 5 by 23.

In [ ]:
# COMPLETE THIS CODE

#### **Problem #4.1.3**

Write code that outputs the result of raising 2 to the 100th power.

In [ ]:
# COMPLETE THIS CODE

#### **Problem #4.1.4**
Write code that creates and outputs a variable called `popular_name` and assign to it the `string` `'Olivia'`.

In [ ]:
# COMPLETE THIS CODE

#### **Problem #4.1.5**

Create a variable `birth_year` and set it equal to the year your were born. Then print, `"I was born in <birth_year>"`.

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

#### **Problem #4.1.6**

Create a variable `favorite_food` and assign to it a `string` with your favorite food. Print `'I love to eat <favorite_food>'`.

In [ ]:
# COMPLETE THIS CODE

In [ ]:
# COMPLETE THIS CODE

#### **Problem #4.1.7**

Debug the code shown below.

In [ ]:
my_list = [10].append(20).append(30)
my_list

AttributeError: ignored

#### **Problem #4.1.8**

Debug the following code.

**HINT**: How many roots should this polynomial have? Even if you're not sure about the math, you could use some of the code below to find out!

In [ ]:
#f(x) = x^2 + 2x + 1
root1, root2 = dynamic_find_roots(1,2,1)
root1

NameError: ignored

In [ ]:
root2

<a name="p4.2"></a>

---
### **Part 4.2: Coding Challenges**
---

This part contains more challenging problems that you can use to push your Python skills to the next level! **Remember what we have learned about the core programming skills of debugging and reading documentation to tackle these problems.**

#### **Problem #4.2.1**

**Independently,** write code that prints out a reversed list.

In [ ]:
my_list = ["one","two","three", "four","five"]
# COMPLETE THIS CODE

#### **Problem #4.2.2**

**Independently,** write code that counts how many times [this article](https://www.ted.com/read/ted-studies/physics/introductory-essay) uses the word "physics".

**Hint:** The string "physics" is different from the string "Physics". You will need to check for both!

In [ ]:
speech = """To the outside observer, it may seem that physics is in some ways the opposite of art and that physicists must sacrifice their artistic intelligence to make way for cold rationality and logic. But nothing could be farther from the truth: Each step forward in our understanding of the universe could not have been conceived without an enormous dose of intuition and creativity. Physicists are on a quest to figure out how nature works at the most fundamental level. This is a romantic story, penned in what may seem the least emotive of languages: mathematics. What's surprising is that the immeasurable beauty of the world is far from lost once its inner workings are expressed in this abstract language. Moreover, there remains something deeply intriguing about the fact that the universe is governed by the rules of mathematics in the first place. As we'll hear from Murray Gell-Mann in the first of the TEDTalks in The Edge of Knowledge, these beautiful mathematical laws are "not merely a conceit of the human mind" — instead, they're an intrinsic part of nature. Many successful ideas in science can be described as beautiful and very often this is a reference to the simplicity and conciseness of nature's laws. Einstein's special and general theories of relativity, which describe how space, time and gravity behave, are based on only three brief postulates. The laws of electromagnetism, which govern every aspect of how we experience the worldthrough sight, sound, smell, taste or touch, are so concise thatthey can be written on the front of a T-shirt. The Standard Model of Particle Physics, which describes all of the known particles and three of the four forces that act between them, fits on the side of a coffee mug. As we will hear from Garrett Lisi, looking for beauty in the patterns that emerge in the laws of physics can tell us about how the universe works at the most fundamental level. Science is a collaborative discipline and a global one too. It is the extent to which scientists cooperate that allows science to move at an incredible pace. The majority of the ideas presented in these TEDTalks have been around no longer than 50 years; some less than a decade. Since these speakers featured in The Edge of Knowledge delivered their TEDTalks, scientists working in global collaborations have developed and implemented several new experimental measurements. Most recently, the European Space Agency's Planck satellite has made precise measurements of the Cosmic Microwave Background (CMB): the results are in agreement with the predictions of the Standard Cosmological Model, which describes how the universe evolved from the Big Bang to what we see today. In Brian Cox's TEDTalk, we'll hear about the search for new elementary particles at CERN's Large Hadron Collider, encompassing the work of over 10,000 physicists from over 100 countries. This search is underway, and appears already to have yielded one of the most important scientific results of the 21st century: the discovery of the Higgs boson, the final ingredient predicted by the Standard Model of Particle Physics. Notwithstanding the significance of these recent discoveries and their agreement with predictions, our picture of the fundamental structure of the universe is far from complete: a number of big mysteries remain in both particle physics and cosmology. As we'll hear from Patricia Burchat, many of these mysteries link together the physics of the smallest elementary particles and the largest distances of the cosmos. One of the most enduring mysteries is how to reconcile a complete theory of gravity with our understanding of the fundamental particles. From Brian Greene, we'll hear about the potential of string theory to solve this problem and the possible existence of tiny, curled up, extra spatial dimensions. However, fundamental laws are not enough on their own. Aristotle said that "all human actions have one or more of these seven causes: chance, nature, compulsion, habit, reason, passion, and desire." It's the first of these — chance — that is not decided by the laws of physics; in fact, chance is not decided at all. The fundamental laws of physics cannot predict what will happen; they can only tell us what might happen. This uncertainty is built into the laws of quantum mechanics. As we'll hear from Aaron O'Connell, the most striking feature of quantum mechanics is that it's weird. For example, we're challenged to contemplate the possibility that a thing can be in more than one place at the same time. It's quantum mechanics, more than any other idea in fundamental physics, which forces us to question our intuition about how everyday objects behave. For the microscopic constituents of the universe, our everyday observations simply do not hold. In spite of its counter-intuitiveness, quantum mechanics has come to define our modern world through the technologies that it underpins. From the tiny switches crammed by the billions onto microchips to medical scanners and laser therapies, all rely upon the weirdness of quantum mechanics. Ultimately, science remains an empirical discipline. Thinking up beautiful theories is not enough on its own: every theory must stand up to the experimental observations of how nature actually works. If it doesn't, then the theory can't be correct and we must try again. If, on the other hand, our observations and predictions agree, then we're encouraged and — if the evidence is sufficient — we might even dare to claim some measure of understanding. Through our human creativity, expressed in a process of trial and improvement, incremental advances in our understanding accumulate and scientific progress is made. As the chess grandmaster Gary Kasparov puts it, our success is "the ability to combine creativity and calculation...into a whole that is much greater than the sum of its parts." With each of these steps forward, science closes one more door and moves on to try one of the many, many doors that remain open. This series of TEDTalks discusses some of the toughest questions and the most profound ideas in fundamental physics. The concepts not only challenge us to think objectively and rationally, but also require us to put aside many of our everyday preconceptions and intuitions about how nature works. Be prepared to re-watch the talks and re-read the supporting material; trying to get your head around 13.8 billion years of the universe's history isn't something you can do in an afternoon! Let's begin with CalTech physicist Murray Gell-Mann for an introduction to the Standard Model of particle physics and the quest for a unified theory."""
# COMPLETE THIS CODE

#### **Problem #4.2.3**

**Independently**, research the `dir()` function and what it does to different objects. Use this function to win the plant game from Part 2!

In [ ]:
# COMPLETE THIS CODE

#### **Problem #4.2.4**

**Independently,** research the `numpy` module to find a function that allows you simulate rolling a dice (picking a random number between 1 and 6).

In [ ]:
# COMPLETE THIS CODE

#### **Problem #4.2.5**

**Independently,** research the `matplotlib` module to find a function that creates a line graph of the data provided below. For an extra challenge, label the x-axis "Year" and the y-axis "Number of Qubits in IBM's Computers". Note that the numbers for 2024 and 2025 are goals set by IBM.

In [ ]:
years = [2019,2020,2021,2022,2023,2024,2025]
num_qubits = [27,65,127,433,1121,1386,4158]
# COMPLETE THIS CODE

#### **Problem #4.2.6**

**Independently,** research the `numpy` library and `pyplot` sub-module of the `matplotlib` library and use its objects, methods, and functions to plot the equation below on a graph whose x-axis ranges from 0 to 42 and whose y-axis ranges from -6 to 6. $$y=5cos(3x)$$

In [ ]:
# COMPLETE THIS CODE

<a name="p5"></a>

---
## **Part 5: The `print` Function [OPTIONAL]**
---

In this part, we will explore how to use functions with the very useful built-in `print` function.

#### **Problem #5.1**

**Together**, let's run the code below to print the text `"Hello Quantum World!"`.

In [ ]:
print("Hello Quantum World!")

Hello Quantum World!


#### **Problem #5.2**

**Together**, let's print the text `"Hello Quantum World!"` and `"Nice to meet you and not meet you"` using two separate print statements.

In [ ]:
print("# COMPLETE THIS CODE )
print("# COMPLETE THIS CODE )

#### **Problem #5.3**

**Independently**, complete the code by printing your age.

In [ ]:
print("My age is: # COMPLETE THIS CODE

#### **Problem #5.4**

**Independently**, Complete the code by printing your age on the next line.

In [ ]:
print("My age is:")
# COMPLETE THIS CODE

#### **Problem #5.5**

Write code to print your favorite animal.

#End of notebook
---
© 2023 The Coding School, All rights reserved